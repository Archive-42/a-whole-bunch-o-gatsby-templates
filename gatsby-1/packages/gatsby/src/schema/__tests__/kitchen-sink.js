// @flow

const { SchemaComposer } = require(`graphql-compose`)
const {
  graphql,
  GraphQLSchema,
  GraphQLNonNull,
  GraphQLList,
  GraphQLObjectType,
  getNamedType,
} = require(`graphql`)
const { store } = require(`../../redux`)
const { actions } = require(`../../redux/actions`)
const { build } = require(`../index`)
const fs = require(`fs-extra`)
const path = require(`path`)
const { slash } = require(`gatsby-core-utils`)
const withResolverContext = require(`../context`)

jest.mock(`../../utils/api-runner-node`)
const apiRunnerNode = require(`../../utils/api-runner-node`)

jest.mock(`gatsby-cli/lib/reporter`, () => {
  return {
    log: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
    activityTimer: () => {
      return {
        start: jest.fn(),
        setStatus: jest.fn(),
        end: jest.fn(),
      }
    },
    phantomActivity: () => {
      return {
        start: jest.fn(),
        end: jest.fn(),
      }
    },
  }
})

// XXX(freiksenet): Expand
describe(`Kitchen sink schema test`, () => {
  let schema
  let schemaComposer

  const runQuery = query =>
    graphql(
      schema,
      query,
      undefined,
      withResolverContext({
        schema,
        schemaComposer,
        context: {},
        customContext: {},
      })
    )

  beforeAll(async () => {
    apiRunnerNode.mockImplementation((api, ...args) => {
      if (api === `setFieldsOnGraphQLNodeType`) {
        return mockSetFieldsOnGraphQLNodeType(...args)
      } else if (api === `createResolvers`) {
        return mockCreateResolvers(...args)
      } else {
        return []
      }
    })

    const nodes = JSON.parse(
      fs
        .readFileSync(
          path.join(__dirname, `./fixtures/kitchen-sink.json`),
          `utf-8`
        )
        .replace(/<PROJECT_ROOT>/g, slash(process.cwd()))
    )

    store.dispatch({ type: `DELETE_CACHE` })
    nodes.forEach(node =>
      actions.createNode(node, { name: `test` })(store.dispatch)
    )

    store.dispatch({
      type: `CREATE_TYPES`,
      payload: `
        interface Post @nodeInterface {
          id: ID!
          code: String
        }
        type PostsJson implements Node & Post @infer {
          id: ID!
          time: Date @dateformat(locale: "fi", formatString: "DD MMMM")
          code: String
          image: File @fileByRelativePath
        }

      `,
    })
    buildThirdPartySchemas().forEach(schema =>
      store.dispatch({
        type: `ADD_THIRD_PARTY_SCHEMA`,
        payload: schema,
      })
    )
    await build({})
    schema = store.getState().schema
    schemaComposer = store.getState().schemaCustomization.composer
  })

  it(`passes kitchen sink query`, async () => {
    expect(
      await runQuery(`
        {
          interface: allPost(sort: { fields: id }, limit: 1) {
            nodes {
              id
              code
              ... on PostsJson {
                time
                image {
                  childImageSharp {
                    id
                  }
                  childrenImageSharp {
                    id
                  }
                }
              }
            }
          }
          sort: allPostsJson(sort: { fields: likes, order: ASC }, limit: 2) {
            edges {
              node {
                id
                idWithDecoration
                time(formatString: "DD.MM.YYYY")
                localeString: time(locale: "ru")
                localeFormat: time(formatString: "DD MMMM YYYY")
                defaultTime: time
                code
                likes
                comment
                image {
                  childImageSharp {
                    id
                  }
                  childrenImageSharp {
                    id
                  }
                }
                _3invalidKey
              }
            }
          }
          filter: allPostsJson(filter: { likes: { eq: null } }, limit: 2) {
            edges {
              node {
                id
                comment
              }
            }
          }
          resolveFilter: postsJson(
            idWithDecoration: { eq: "decoration-1601601194425654597" }
          ) {
            id
            idWithDecoration
            likes
          }
          createResolvers: likedEnough {
            id
            likes
            code
          }
          thirdPartyStuff {
            text
            child {
              ... on ThirdPartyStuff {
                text
              }
              ... on ThirdPartyStuff2 {
                foo
              }
            }
          }
          thirdPartyUnion {
            ... on ThirdPartyStuff {
              text
            }
            ... on ThirdPartyStuff2 {
              foo
            }
          }
          thirdPartyInterface {
            ... on ThirdPartyStuff3 {
              text
            }
          }
        }
      `)
    ).toMatchSnapshot()
  }, 30000)

  it(`correctly resolves nested Query types from third-party types`, () => {
    const queryFields = schema.getQueryType().getFields()
    ;[`relay`, `relay2`, `query`, `manyQueries`].forEach(fieldName =>
      expect(getNamedType(queryFields[fieldName].type)).toBe(
        schema.getQueryType()
      )
    )
    expect(schema.getType(`Nested`).getFields().query.type).toBe(
      schema.getQueryType()
    )
  })
})

const buildThirdPartySchemas = () => {
  const schemaComposer = new SchemaComposer()
  schemaComposer.addTypeDefs(`
    type ThirdPartyStuff {
      text: String
      child: ThirdPartyUnion2
    }

    type ThirdPartyStuff2 {
      foo: String
    }

    union ThirdPartyUnion = ThirdPartyStuff | ThirdPartyStuff2

    interface ThirdPartyInterface {
      text: String
      relay: Query
    }

    type ThirdPartyStuff3 implements ThirdPartyInterface {
      text: String
      relay: Query
    }

    union ThirdPartyUnion2 = ThirdPartyStuff | ThirdPartyStuff2

    type Query {
      thirdPartyStuff: ThirdPartyStuff
      thirdPartyUnion: ThirdPartyUnion
      thirdPartyInterface: ThirdPartyInterface
      relay: Query
      relay2: [Query]!
    }
  `)
  schemaComposer
    .getUTC(`ThirdPartyUnion`)
    .setResolveType(() => `ThirdPartyStuff`)
  schemaComposer
    .getUTC(`ThirdPartyUnion2`)
    .setResolveType(() => `ThirdPartyStuff`)
  schemaComposer
    .getIFTC(`ThirdPartyInterface`)
    .setResolveType(() => `ThirdPartyStuff3`)
  schemaComposer.Query.extendField(`thirdPartyStuff`, {
    resolve() {
      return {
        text: `Hello third-party schema!`,
        child: {
          text: `Hello from children!`,
        },
      }
    },
  })
  schemaComposer.Query.extendField(`thirdPartyUnion`, {
    resolve() {
      return {
        text: `Hello third-party schema!`,
        child: {
          text: `Hello from children!`,
        },
      }
    },
  })
  schemaComposer.Query.extendField(`thirdPartyInterface`, {
    resolve() {
      return {
        text: `Hello third-party schema!`,
      }
    },
  })
  schemaComposer.addSchemaMustHaveType(
    schemaComposer.getOTC(`ThirdPartyStuff3`)
  )

  // Query type with non-default name
  const RootQueryType = new GraphQLObjectType({
    name: `RootQueryType`,
    fields: () => {
      return {
        query: { type: RootQueryType },
        manyQueries: {
          type: new GraphQLNonNull(new GraphQLList(RootQueryType)),
        },
        nested: { type: Nested },
      }
    },
  })
  const Nested = new GraphQLObjectType({
    name: `Nested`,
    fields: () => {
      return {
        query: { type: RootQueryType },
      }
    },
  })
  const schema = new GraphQLSchema({ query: RootQueryType })

  return [schemaComposer.buildSchema(), schema]
}

const mockSetFieldsOnGraphQLNodeType = async ({ type: { name } }) => {
  if (name === `PostsJson`) {
    return [
      {
        idWithDecoration: {
          type: `String`,
          resolve(parent) {
            return `decoration-${parent.id}`
          },
        },
      },
    ]
  } else {
    return []
  }
}

const mockCreateResolvers = ({ createResolvers }) => {
  createResolvers({
    Query: {
      likedEnough: {
        type: `[PostsJson]`,
        async resolve(parent, args, context) {
          const result = await context.nodeModel.runQuery({
            type: `PostsJson`,
            query: {
              filter: {
                likes: {
                  ne: null,
                  gt: 5,
                },
              },
              sort: {
                fields: [`likes`],
                order: [`DESC`],
              },
            },
            firstOnly: false,
          })
          return result.slice(0, 2)
        },
      },
    },
  })
}
