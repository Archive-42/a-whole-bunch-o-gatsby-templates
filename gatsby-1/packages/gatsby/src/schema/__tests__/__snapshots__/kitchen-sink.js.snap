// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kitchen sink schema test passes kitchen sink query 1`] = `
Object {
  "data": Object {
    "createResolvers": Array [
      Object {
        "code": "BdiU-TTFP4h",
        "id": "1685001452849004065",
        "likes": 33,
      },
      Object {
        "code": "Bam3G5XFOTK",
        "id": "1632234281433883850",
        "likes": 30,
      },
    ],
    "filter": Object {
      "edges": Array [
        Object {
          "node": Object {
            "comment": 1,
            "id": "1270677182602272387",
          },
        },
        Object {
          "node": Object {
            "comment": 0,
            "id": "1256134251849702933",
          },
        },
      ],
    },
    "interface": Object {
      "nodes": Array [
        Object {
          "code": "3lADm0M90E",
          "id": "1001206739996237060",
          "image": Object {
            "childImageSharp": Object {
              "id": "0b06bdfe-f253-5859-8321-b42b5d737195",
            },
            "childrenImageSharp": Array [
              Object {
                "id": "0b06bdfe-f253-5859-8321-b42b5d737195",
              },
            ],
          },
          "time": "06 kesäkuu",
        },
      ],
    },
    "resolveFilter": Object {
      "id": "1601601194425654597",
      "idWithDecoration": "decoration-1601601194425654597",
      "likes": 9,
    },
    "sort": Object {
      "edges": Array [
        Object {
          "node": Object {
            "_3invalidKey": null,
            "code": "BShF_8qhtEv",
            "comment": 0,
            "defaultTime": "05 huhtikuu",
            "id": "1486495736706552111",
            "idWithDecoration": "decoration-1486495736706552111",
            "image": Object {
              "childImageSharp": Object {
                "id": "f5f0e564-899f-5bbc-92de-a2325b69cb75",
              },
              "childrenImageSharp": Array [
                Object {
                  "id": "f5f0e564-899f-5bbc-92de-a2325b69cb75",
                },
              ],
            },
            "likes": 8,
            "localeFormat": "05 huhtikuu 2017",
            "localeString": "05 апреля",
            "time": "05.04.2017",
          },
        },
        Object {
          "node": Object {
            "_3invalidKey": null,
            "code": "BY6B8z5lR1F",
            "comment": 0,
            "defaultTime": "11 syyskuu",
            "id": "1601601194425654597",
            "idWithDecoration": "decoration-1601601194425654597",
            "image": Object {
              "childImageSharp": Object {
                "id": "6ba97198-5331-53d6-a0e0-c9c063c5094e",
              },
              "childrenImageSharp": Array [
                Object {
                  "id": "6ba97198-5331-53d6-a0e0-c9c063c5094e",
                },
              ],
            },
            "likes": 9,
            "localeFormat": "11 syyskuu 2017",
            "localeString": "11 сентября",
            "time": "11.09.2017",
          },
        },
      ],
    },
    "thirdPartyInterface": Object {
      "text": "Hello third-party schema!",
    },
    "thirdPartyStuff": Object {
      "child": Object {
        "text": "Hello from children!",
      },
      "text": "Hello third-party schema!",
    },
    "thirdPartyUnion": Object {
      "text": "Hello third-party schema!",
    },
  },
}
`;
